api/service/stagedsync/stage_bodies.go

```diff
@@ -0,0 +1,697 @@
+package stagedsync
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/queue"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon-lib/kv/mdbx"
+	"github.com/ledgerwatch/log/v3"
+)
+
+type StageBodies struct {
+	configs StageBodiesCfg
+}
+type StageBodiesCfg struct {
+	ctx           context.Context
+	bc            core.BlockChain
+	db            kv.RwDB
+	turbo         bool
+	turboModeCh   chan struct{}
+	bgProcRunning bool
+	isBeacon      bool
+	cachedb       kv.RwDB
+}
+
+func NewStageBodies(cfg StageBodiesCfg) *StageBodies {
+	return &StageBodies{
+		configs: cfg,
+	}
+}
+
+func NewStageBodiesCfg(ctx context.Context, bc core.BlockChain, db kv.RwDB, isBeacon bool, turbo bool) StageBodiesCfg {
+	cachedb, err := initBlocksCacheDB(ctx, isBeacon)
+	if err != nil {
+		panic("can't initialize sync caches")
+	}
+	return StageBodiesCfg{
+		ctx:      ctx,
+		bc:       bc,
+		db:       db,
+		turbo:    turbo,
+		isBeacon: isBeacon,
+		cachedb:  cachedb,
+	}
+}
+
+func initBlocksCacheDB(ctx context.Context, isBeacon bool) (db kv.RwDB, err error) {
+	// create caches db
+	cachedbName := Block_Cache_DB
+	if isBeacon {
+		cachedbName = "beacon_" + cachedbName
+	}
+	cachedb := mdbx.NewMDBX(log.New()).Path(cachedbName).MustOpen()
```

Comment:
So every stage has it's own DB?

Reply:
Every stage uses main sync db (either memory or disk) and also every stage can has its own cache db (if needed) and then it clears it in prune function (if needed).